### ISSUE TITLE
Query re functioning of rollup building

### ISSUE THREAD

**Issue by aardgoose** [mrdoob/three.js#9584]
##### Description of the problem

I have assembled an application using ES6 modules including three.js as modules using rollup (latest version).  If I attempt to import a single class, eg Vector3(), from 'src/Three.js'
the resultant file has all the three.js modules.

eg: 

_import  { Vector3 } from '../three.js/src/Three.js';
var v = new Vector3();
console.log( v );_

 Is this the expected behaviour, a problem with the modulization or a problem with rollup's treeshaking?

I know I can import the individual modules src/Math/Vector3.js etc, but this is fairly tedious/error prone, so I would prefer to avoid it if possible.

On the plus side rollup makes for a much nicer build path for my app.
##### Three.js version
- [ ] Dev
- [X] r80
- [ ] ...
##### Browser
- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer
##### OS
- [x] All of them
- [ ] Windows
- [ ] Linux
- [ ] Android
- [ ] IOS

**Comment by GGAlanSmithee** [comment]
For rollup to handel your usecase (tree shaking) we need to add a (working) `jsnext:main` entry point. See #9402 forbsome additional info.

**Comment by aardgoose** [comment]
Thanks,

I'm not sure I follow why treeshaking is only possible from a monolithic output from a jsnext target rather that from the src/Three.js file that is used to build  it, and has the same information about dependencies available to it.  Its not a big issue, and happy to live with the current situation, certainly better than my cobbled together build arrangement. I can see the concerns with glsl files, but then I have my own so I'm using the same rollup plugin as Three.

**Comment by GGAlanSmithee** [comment]
Sorry, I missundersrood you. I thought you imported three from npm, as a bundle. It is actually a bit weird that it imports everything in your case.. I will try to investigate, but it almost sounds as an issue with rollup or your setup, because I don't think three does anything out of the ordinary which would prevent three shaking (other than some special handling of geometries). Could you share you rollup config?

**Comment by aardgoose** [comment]
Certainly, test case files attached below, (renamed from .js to allow attaching)

rollup v 0.34.10
node v 4.4.7

Windows 10 latest update.


[rollup-demo.txt](https://github.com/mrdoob/three.js/files/440583/rollup-demo.txt)
[demo.txt](https://github.com/mrdoob/three.js/files/440584/demo.txt)

**Comment by cecilemuller** [comment]
Three's rollup config could use `targets` to build a ES bundle as well, to avoid having to import from the src file to use tree shaking, for example:

````javascript
export default {
	entry: 'src/Three.js',
	plugins: [
		glsl()
	],
	targets: [
		{
			format: 'umd',
			moduleName: 'THREE',
			dest: 'build/three.js'
		},
		{
			format: 'es',
			dest: 'build/three.jsnext.js'
		}
	],
	outro: outro
};
````

instead of currently:

````javascript
export default {
	entry: 'src/Three.js',
	dest: 'build/three.js',
	moduleName: 'THREE',
	format: 'umd',
	plugins: [
		glsl()
	],
	outro: outro
};
````

Although when I had a quick try yesterday in a fork and copied the outputted esnext bundle ~930Kb ([three.esnext.zip](https://github.com/mrdoob/three.js/files/440590/three.esnext.zip)) to a test app that only imported Vector3, it still ended up with a large output after tree shaking (~820Kb).

**Comment by GGAlanSmithee** [comment]
@cecilemuller Nice, it looks as if your setup actually solves #9402. There are probably lots of optimizations which would make the underwhelming results vetter.

**Comment by cecilemuller** [comment]
I re-forked (had deleted it because I was only testing Rollup the other day to see if it's worth converting my webpack stacks), [added the two targets](https://github.com/wildpeaks/three.js/blob/dev/rollup.config.js) (`umd` and `es`) and rebuilt.

Oddly enough, the regular `umd` bundle is ~20Kb smaller (probably because the latest version of dependencies have some new optimizations):
https://github.com/wildpeaks/three.js/tree/dev/build


Then also uploaded a test app (just using Vector3) that compares importing from src and from the bundles:
https://github.com/wildpeaks/example-rollup-threejs-import
(for building from src, I had to copy the `glsl` function from [three's rollup.config.js](https://github.com/mrdoob/three.js/blob/dev/rollup.config.js)).


The result is they're all about 687Kb, and importing from src is about 1Kb smaller than importing from either the `es` or `umd` bundles.

**Comment by cecilemuller** [comment]
Also just added a test app that imports Vector3 from `/src/math/Vector3.js` and wow, this time the tree shaking managed to remove a lot more, the output is barely 36Kb :smiley:

**Comment by vanruesc** [comment]
Hi, I encountered a similar problem a while ago.

> Although I do wonder why it doesn't have the same result when importing it from the ES Bundle or even /src/Three.js considering it's merely forwarding the Vector3 export.

That's because you kept the other files of Three a secret from Rollup. If you import from the entry file `src/Three.js`, Rollup will be aware of all the other existing files of which quite many contain certain statements that may have side-effects.

The tree-shaking mechanism is rather conservative and statements such as `Object.assign` will make it fail immediately, resulting in the inclusion of the file it occured in.

A related Rollup discussion: https://github.com/rollup/rollup/issues/551

**Comment by GGAlanSmithee** [comment]
also, it makes sense that the `Vector3` import yields an output that is a lot smaller since that class only depends on:

```javascript
import { _Math } from './Math';
import { Matrix4 } from './Matrix4';
import { Quaternion } from './Quaternion';
```

So as long as we get the `esnext:main` bundle working, we can start optimiznig away things that prevents treeshaking.

**Comment by cecilemuller** [comment]
@vanruesc If you look at the source, you'll see that the test app does only imports Vector3 from Three.js (which is itself just a export forwarding Vector3.js).

And yes, Vector3 depending on only 3 other simple files is exactly why I picked it because it's meant to make a stark difference in filesize when shaking works :smiley: 

The issue is that Rollup doesn't identify `import {Vector3} from 'somefile.js'` as meaning "only Vector3 is used", hence why I'm curious what exactly is preventing it from identifying it.

---

Replacing the contents of `/src/Three.js` by this is enough to reproduce the issue (Box2 is not removed):

````javascript
export {Box2} from './math/Box2.js';
export {Vector3} from './math/Vector3.js';
````

Yet Vector2 gets correctly removed with this one instead:

````javascript
export {Vector2} from './math/Vector2.js';
export {Vector3} from './math/Vector3.js';
````

Vector2 has no dependency, Box2 only depends on Vector2, they don't even have any dependency in common with Vector3, and Box2 has no cyclical dependency either, so it reduces the scope of what could be causing it.

**Comment by cecilemuller** [comment]
I kept simplifying Box2 until I could find what exactly prevents it from detecting, and it seems a **self-executing function** is the cause.


If Box2.js contents are replaced by this, Box2 is not removed:

````javascript
import {Vector2} from './Vector2';

function Box2(){}

Box2.prototype = {
	setFromCenterAndSize: function(){
	 	var v1 = new Vector2();
	}()
};

export { Box2 };
````

Whereas if it's the following, Box2 is correctly identified as not used and removed:

````javascript
import {Vector2} from './Vector2';

function Box2(){}

Box2.prototype = {
	setFromCenterAndSize: function(){
		var v1 = new Vector2();
	}
};

export { Box2 };
````

**Comment by GGAlanSmithee** [comment]
@cecilemuller I was also trying to find the cause, and I do not think it is the self-executing function itself that is the cause, but rather the instantiation on the ´v1´ variable outside the closure. For example, this will make tree-shaking work as intended:

´´´javascript
setFromCenterAndSize: function () {

		return function setFromCenterAndSize( center, size ) {
			
			var v1 = new Vector2();
			
			var halfSize = v1.copy( size ).multiplyScalar( 0.5 );
			this.min.copy( center ).sub( halfSize );
			this.max.copy( center ).add( halfSize );

			return this;

		};

	}(),
´´´

and the self-execution could then be removed ofc.. This is however de-optimizing the funciton in terms of speed, which I guess is the reason for instantiating the ´v1´ variable outside the closure in the first place.

I do not know why this is preventing ´rollup´ from tree-shaking it though..

**Comment by vanruesc** [comment]
> If you look at the source, you'll see that the test app does only imports Vector3 from /src/Three.js (which is itself just a export forwarding Vector3.js).

Yes, but Rollup has to process the export catalog `src/Three.js` to reach the desired dependency and that alone exposes the other files to the bundler.

Apparently, Rollup can't easily determine whether the immediate instantiation of `v1` inside the iife is side-effect-free. _The tree-shaking bails out fast_, probably for bundling-performance reasons. If the instantiation is inside a function declaration, it's considered safe.

The bewildering thing about this is that `Box2` gets included in the final bundle although it's not used by `Vector3`. In my opinion, the unrelated file shouldn't even be visited by Rollup, but, _to be safe_, it does just that and includes it regardless because of the questionable instantiation.

> I do not know why this is preventing ´rollup´ from tree-shaking it though

As far as I know, improving the logic for side-effect detection hasn't been the development focus of Rollup thus far. I can imagine that the current behaviour is due to performance considerations.

**Comment by aardgoose** [comment]
Is there a definitive list of things that rollup defines as having possible side effects?

This discussion appears to have identified Object.assign, Object.definePropert(ies|y) and instantiation with iife, all of which are common throughout Three.js.

Removing Three.Legacy.js ( basically a lot of Object.assigns ) from the exports from Three.js shaves about 230K from the simple Vector example which is a start. The question now is what changes are worth doing to remove "side effects", and what is the roadmap for rollup for more sophisticated analysis.

Re the visitation of Box2, I suspect, in effect Vector3 AND Box2 are imported before being re-exported into the abbreviated Three.js, and it's in the context of Three.js that Box2 is evaluated for potential side effects on other items within the Tree.js module, rather than in the context of the entrypoint .js. Pure speculation on my part, new to ES6.

As a test I made all the exports in Three.js into "imports" and then made a single export for Vector3, and still had the same bloated output, which suggests something similar to the above is happening.

Anyway, an interesting discussion, and thanks for confirming I am not making a simple mistake somewhere.

**Comment by vanruesc** [comment]
> Is there a definitive list of things that rollup defines as having possible side effects?

Unfortunately not. You'd have to check Rollup's [sources](https://github.com/rollup/rollup/blob/master/src) (Module, Statement, Bundle, ...) to find out which statements cause module inclusions.

> I suspect, in effect Vector3 AND Box2 are imported before being re-exported into the abbreviated Three.js, and it's in the context of Three.js that Box2 is evaluated for potential side effects

I think you could interpret it that way, yes. `Vector3` and `Box2` will implicitly be imported by `src/Three.js` and then re-exported.

**Comment by cecilemuller** [comment]
Out of curiosity, I also tried with Webpack 2 beta (as it claims to have tree-shaking too), but sadly not only it didn't remove Box2, it didn't remove Vector2 either:
https://github.com/wildpeaks/example-webpack2-threejs-import

**Comment by cecilemuller** [comment]
Updated the [Webpack 2 import example](https://github.com/wildpeaks/example-webpack2-threejs-import/) to be more verbose about why/what it chooses to include (using `--display-reasons --display-used-exports`).

**Comment by GGAlanSmithee** [comment]
@cecilemuller 

It would be really nice to re-test the Box2 / Vector2 tree-shaking example with the `rewrite` branch of rollup, as described in https://github.com/rollup/rollup/pull/902. 

If I have time later, I will give it a try, no idea if it will behave differently in our use-case, but might be nice to provide feedback to rollup if nothing else.

**Comment by cecilemuller** [comment]
I tried to, however I'm not certain what's the correct way to make package.json depend on a different branch, `npm install` failed for each of those:

	"rollup": "rollup/rollup#rewrite"
	"rollup": "git://github.com/rollup/rollup.git#rewrite",
	"rollup": "git+ssh://git@github.com:rollup/rollup.git#rewrite"
	"rollup": "git+https://git@github.com/rollup/rollup.git#rewrite"
	"rollup": "https://git@github.com/rollup/rollup.git#rewrite"

The only thing that worked was manually cloning the branch, npm install in the branch, then use a relative path:

	"rollup": "file:../rollup/"


--

However it fails to build the test apps at the moment, each test gives a parse error (although not all have the same error):

````text

rollup --config ./rollup.import-src.js

	Error parsing D:\Github\example-rollup-threejs-import\threejs\src\core\EventDispatcher.js: Class constructor DestructuringErrors cannot be invoked without 'new' in D:\Github\example-rollup-threejs-import\threejs\src\core\EventDispatcher.js


rollup --config ./rollup.import-src-vector3.js

	Cannot read property 'add' of undefined
	TypeError: Cannot read property 'add' of undefined
		at UpdateExpression.bind (D:\Github\example-rollup-threejs-import\node_modules\rollup\src\ast\nodes\UpdateExpression.js:17:30)
		at eachChild.child (D:\Github\example-rollup-threejs-import\node_modules\rollup\src\ast\Node.js:6:34)
		at ExpressionStatement.eachChild (D:\Github\example-rollup-threejs-import\node_modules\rollup\src\ast\Node.js:21:6)
		at ExpressionStatement.bind (D:\Github\example-rollup-threejs-import\node_modules\rollup\src\ast\Node.js:6:8)
		at BlockStatement.bind (D:\Github\example-rollup-threejs-import\node_modules\rollup\src\ast\nodes\BlockStatement.js:8:9)
		at FunctionExpression.bind (D:\Github\example-rollup-threejs-import\node_modules\rollup\src\ast\nodes\FunctionExpression.js:7:13)
		at eachChild.child (D:\Github\example-rollup-threejs-import\node_modules\rollup\src\ast\Node.js:6:34)
		at Node$1.eachChild (D:\Github\example-rollup-threejs-import\node_modules\rollup\src\ast\Node.js:21:6)
		at Node$1.bind (D:\Github\example-rollup-threejs-import\node_modules\rollup\src\ast\Node.js:6:8)
		at eachChild.child (D:\Github\example-rollup-threejs-import\node_modules\rollup\src\ast\Node.js:6:34)


rollup --config ./rollup.import-build-esnext.js

	Error parsing D:\Github\example-rollup-threejs-import\threejs\build\three.esnext.js: Class constructor DestructuringErrors cannot be invoked without 'new' in D:\Github\example-rollup-threejs-import\threejs\build\three.esnext.js


rollup --config ./rollup.import-build-regular.js

	Error parsing D:\Github\example-rollup-threejs-import\threejs\build\three.esnext.js: Class constructor DestructuringErrors cannot be invoked without 'new' in D:\Github\example-rollup-threejs-import\threejs\build\three.esnext.js
````

**Comment by aardgoose** [comment]
I also built from a git pull and encountered similar (but not identical) errors in EventDispatcher FWIW.
ie I got: "Class constructors cannot be invoked without 'new'" when processing a a single import of Vector3 from src/Three.js.

Previously I also looked at instrumenting side effect detection in 
rollup/src/utils/run.js which appears to be where potential side effects are flagged, which may prove useful in tracking down reasons for inclusions.  Functions with no side effects are in rollup/src/utils/pureFunctions.js, with Object.assign notable by its omission.

The build self tests break with the added warnings, presumably because of unexpected output.

( I assume you know the rollup code more that I do, so you probably know this already - I just scanned through the code briefly).

**Comment by GGAlanSmithee** [comment]
Sorry, I did not have time to test myself today, but looks like the `class constructors cannot be invoked without new` error was encountered and fixed within rollup/rollup#902 if you would like to give it another try.

**Comment by GGAlanSmithee** [comment]
Sorry @aardgoose, I'm not too familiar with the internals of rollup. My guess is that `Object.assign()` is considered impure since it invokes getters / getters wich in turn might have sideceffects: 

> It uses [[Get]] on the source and [[Set]] on the target, so it will invoke getters and setters

[source](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign)

**Comment by cecilemuller** [comment]
@GGAlanSmithee Just retried with fresh installs and unfortunately it's still outputting the same errors.

**Comment by GGAlanSmithee** [comment]
@cecilemuller 

`rollup` has published version `^0.35`, and it seems like it works to build your tests again. I sent you a PR that adds a build config for `box2` as well as a modified version of it, `box2-modified`, which is identical with the expection of having the `Vector2` instatiation inside the closures of the `distanceToPoint` and `setFromCenterAndSize` functions.

It works to build both `box2` and `box2-modified` but unfortunatley `rollup` does not tree-shake _even the modified version_. This is unexpected, since it did so in version `^0.34`.

@Rich-Harris if you want to take a look at this (since it could be a regression), here are some steps:

1. clone https://github.com/wildpeaks/example-rollup-threejs-import (`rewrite` branch after my PR is accepted or look at my fork).
2. run `npm i && npm run build`
3. compare `dist/bundle.import-src-box2es6.js` and `dist/bundle.import-src-box2-modified.es6.js`
    * expected modified version to tree-shake all code, but does not
4. run `npm uninstall rollup -D && npm i rollup@^0.34 -D && npm run build`
5. compare `dist/bundle.import-src-box2es6.js` and `dist/bundle.import-src-box2-modified.es6.js` 
    * now, `dist/bundle.import-src-box2-modified.es6.js` is empty (except the `console.log`)

This is a bit of a tangent since the original question was why the closure (or self-executing function) causes rollup to not tree-shake at all, so if you want to answer that, be my guest :+1: 

Also @aardgoose regargin `Object.` methods not being tree-shaken, see https://github.com/rollup/rollup/pull/555 for some additional info

**Comment by aardgoose** [comment]
Closing this, I think my initial question was answered a while ago.
